### PR TITLE
[gen] reject generation if -relax cannot provide cycles

### PR DESCRIPTION
**Pull request by hugookeeffe**
This patch rejects generation of cycle when -safe and -relax are used in a configuration file, but relaxations in -relax cannot generate cycles.

Currently, the generator still generates cycles from the safe list if this is the case.
an example of where this can occur is:
`diy7 -cycleonly true -arch AArch64 -safe "DMB.SYd** Rfe Fre Coe" -relax "[PodRR,PodWR]"`

The reason for this patch is that in our current flow, some composite relaxations are automatically generated to go into `-relax`, Occasionally junk relaxations are accidentally created (e.g. [PodRR, PodWR]), which cannot provide cycles, however the generator will still use relaxations in -safe. 

This patch introduces a fail safe to stop this behaviour.

### CODE REVIEW

**Comment by hugookeeffe** [comment]
cc: @relokin @jalglave @maranget @chacon01

**Comment by maranget** [comment]
Hi @hugookeeffe. You are right no cycle should be generated in the situation you describe. As have a slight remark, more of aesthetic order: generally I tend to avoid polymorphic equality as much as possible. I suggest the following cosmetic modification.

File: gen/alt.ml [hunk 1]
```diff
@@ -684,9 +684,13 @@ module Make(C:Builder.S)
       List.iter (fun r -> fprintf chan "%s\n" (pp_relax r)) rs
 
     let secret_gen relax safe reject n =
+      let r_nempty = not (relax = []) in
```
Comment (maranget) [inline gen/alt.ml:687]:
```suggestion
      let r_nempty = Misc.consp relax in
```

File: gen/alt.ml [hunk 2]
```diff
@@ -684,9 +684,13 @@ module Make(C:Builder.S)
...
       let relax = expand_relaxs C.ppo relax
       and safe = expand_relaxs C.ppo safe
       and reject = expand_relaxs C.ppo reject in
+      if relax = [] then if r_nempty then begin
```
Comment (maranget) [inline gen/alt.ml:691]:
```suggestion
      if Misc.nilp relax then if r_nempty then begin
```

**Comment by hugookeeffe** [comment]
I think I have made all the suggested changes, let me know if any others need to be made :)

**Comment by maranget** [comment]
Merged, thanks.

**Pull request merged**